File: tests/binding/test_drp_playhead_invariant.lua
-- Regression: every sequence created by the DRP import path must satisfy
-- `playhead_frame >= start_timecode_frame`. A playhead below the sequence's
-- TC origin trips the playback engine's start-frame assert at seek time,
-- and pre-content space is meaningless anyway. The Sequence.create
-- invariant (models/sequence.lua) enforces this at construction time —
-- this test pins the importer call paths as honoring it end-to-end.
--
-- Fixture: sample_project.drp (small, fast). The larger fixtures
-- (anamnesis-gold-timeline.drp, "2025-06-14 NO KINGS SEATTLE.drp") are
-- verified manually via the same probe pattern but not run in the auto-test
-- suite to keep runtime tight; slow-tier coverage can be added under
-- tests/slow/ if regressions appear with TC-bearing media.

require("test_env")

local Sequence = require("models.sequence")
local database = require("core.database")
local test_env = require("test_env")

print("\n=== DRP Playhead Invariant Test ===")

local fixture_path = test_env.resolve_repo_path(
    "tests/fixtures/resolve/sample_project.drp")
local jvp_path = "/tmp/jve/test_drp_playhead_invariant.jvp"
os.remove(jvp_path)
os.remove(jvp_path .. "-wal")
os.remove(jvp_path .. "-shm")

local ok, err = require("core.commands.open_project")._convert_drp_to_jvp(
    fixture_path, jvp_path, nil, { audio_sample_rate = 48000 })
assert(ok, "convert failed: " .. tostring(err))

local conn = assert(database.get_connection(), "no database connection")
local stmt = assert(conn:prepare("SELECT id FROM sequences"))
assert(stmt:exec())
local ids = {}
while stmt:next() do table.insert(ids, stmt:value(0)) end
stmt:finalize()
assert(#ids > 0, "import produced no sequences")

local masters, timelines = 0, 0
for _, id in ipairs(ids) do
    local s = Sequence.load(id)
    assert(s.playhead_position >= s.start_timecode_frame, string.format(
        "DRP import violates playhead invariant for sequence '%s' (kind=%s): "
        .. "playhead_position=%d < start_timecode_frame=%d — track down "
        .. "which Sequence.create call site emitted a sub-TC-origin playhead",
        s.name, s.kind, s.playhead_position, s.start_timecode_frame))
    if s.kind == "master" then masters = masters + 1
    else timelines = timelines + 1 end
end
print(string.format("  %d masters + %d timelines = %d sequences, all satisfy "
    .. "playhead_position >= start_timecode_frame",
    masters, timelines, #ids))

print("\n✅ test_drp_playhead_invariant.lua passed")

File: tests/smoke/cases/test_drp_playhead_invariant.py
# TODO: needs `core.debug_helpers.sequence_ids()` (and underlying
# `Sequence.list_all_ids()`) to enumerate every sequence produced by a
# DRP import — see MIGRATION_ANALYSIS.md entry for
# tests/binding/test_drp_playhead_invariant.lua. Reads through the
# anamnesis substrate happen to expose a single record sequence via
# `displayed_sequence_id()`, but the invariant must hold for EVERY
# imported sequence (masters + timelines), so a per-id enumerator is
# required. Until then this smoke is skipped.
"""
DRP import invariant — every sequence produced by the DRP→JVP conversion
path satisfies playhead_position >= start_timecode_frame. A playhead
below the sequence's TC origin trips the C++ playback engine's
start-frame assert at seek time, so pre-content space is meaningless
anyway. Sequence.create enforces this at construction; this smoke pins
the importer call paths as honoring it end-to-end.

Origin: tests/binding/test_drp_playhead_invariant.lua (pre-2026-05-30
binding test that called `_convert_drp_to_jvp` directly; replaced here
by driving File > Open Project... on `sample_project.drp` through the
real menu + file dialog).

Run:
    python3 -m unittest tests.smoke.cases.test_drp_playhead_invariant -v
"""

import sys
import unittest
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(REPO_ROOT))

from tests.smoke.runner.case import JVESmokeCase

FIXTURE_DRP = REPO_ROOT / "tests" / "fixtures" / "resolve" / "sample_project.drp"


class TestDRPPlayheadInvariant(JVESmokeCase):
    """Every imported sequence must satisfy playhead >= start_timecode_frame."""

    @unittest.skip("needs core.debug_helpers.sequence_ids() enumerator — "
                   "see TODO at top of file")
    def test_drp_open_emits_no_sub_tc_origin_playhead(self) -> None:
        # Fresh project at class setup is the anamnesis template; for
        # this test we want a pristine open of the DRP fixture so the
        # only sequences in the project are the ones the importer
        # emitted. Reset to template first, then drive File > Open
        # Project... onto the DRP — JVE's open path converts .drp to
        # .jvp on the way in (the behavior we're pinning).
        self._reset_to_template()

        self.assertTrue(FIXTURE_DRP.exists(),
            f"fixture missing: {FIXTURE_DRP}")

        before = self.eval_int(
            'return require("core.debug_helpers").sequence_count()')

        self.menu_pick("File > Open Project...")
        self.pick_file_in_open_dialog(str(FIXTURE_DRP))

        # Import is async — wait for the sequence table to grow past
        # the pre-import count.
        self.wait_for(
            f'return require("core.debug_helpers").sequence_count() > {before}',
            timeout=20.0)

        # Enumerate every sequence id and assert the invariant.
        # `sequence_ids()` is the missing primitive that gates this
        # test — must return a delimited string of all sequence ids in
        # the active project so the smoke can split + iterate.
        ids_blob = self.eval_str(
            'return table.concat('
            '  require("core.debug_helpers").sequence_ids(), "|")')
        ids = [s for s in ids_blob.strip('"').split("|") if s]
        self.assertGreater(len(ids), 0,
            "DRP import produced no sequences — importer broken or "
            "the Open dialog didn't accept the file.")

        for seq_id in ids:
            playhead = self.eval_int(
                f'return require("core.debug_helpers").playhead_of("{seq_id}")')
            start_tc = self.eval_int(
                f'return require("core.debug_helpers").sequence_start_tc("{seq_id}")')
            name = self.eval_str(
                f'return require("core.debug_helpers").sequence_field("{seq_id}", "name")')
            kind = self.eval_str(
                f'return require("core.debug_helpers").sequence_field("{seq_id}", "kind")')
            self.assertGreaterEqual(playhead, start_tc, (
                f"DRP import violates playhead invariant for sequence "
                f"'{name}' (kind={kind}): "
                f"playhead_position={playhead} < "
                f"start_timecode_frame={start_tc} — track down which "
                f"Sequence.create call site emitted a sub-TC-origin "
                f"playhead."))


if __name__ == "__main__":
    unittest.main()
